cpp/ql/lib/semmle/code/cpp/dataflow/internal/FlowSummaryImpl.qll

```diff
@@ -0,0 +1,241 @@
+   * `output`, kind `kind`, and provenance `provenance`.
+   */
+  predicate sourceElement(SourceOrSinkElement e, string output, string kind) {
+    exists(
+      string namespace, string type, boolean subtypes, string name, string signature, string ext
+    |
+      sourceModel(namespace, type, subtypes, name, signature, ext, output, kind, _) and
+      e = interpretElement(namespace, type, subtypes, name, signature, ext)
+    )
+  }
+
+  /**
+   * Holds if an external sink specification exists for `e` with input specification
+   * `input`, kind `kind` and provenance `provenance`.
+   */
+  predicate sinkElement(SourceOrSinkElement e, string input, string kind) {
+    exists(
+      string package, string type, boolean subtypes, string name, string signature, string ext
+    |
+      sinkModel(package, type, subtypes, name, signature, ext, input, kind, _) and
+      e = interpretElement(package, type, subtypes, name, signature, ext)
+    )
+  }
+
+  private newtype TInterpretNode =
+    TElement_(Element n) or
+    TNode_(Node n)
+
+  /** An entity used to interpret a source/sink specification. */
+  class InterpretNode extends TInterpretNode {
+    /** Gets the element that this node corresponds to, if any. */
+    SourceOrSinkElement asElement() { this = TElement_(result) }
+
+    /** Gets the data-flow node that this node corresponds to, if any. */
+    Node asNode() { this = TNode_(result) }
+
+    /** Gets the call that this node corresponds to, if any. */
+    DataFlowCall asCall() {
+      this.asElement() = result.asCallInstruction().getUnconvertedResultExpression()
+      // TODO: or summary call?
+      // ??? this.asNode().(FlowSummaryNode).getSummaryNode() = result.(SummaryCall).getReceiver()
+    }
+
+    /** Gets the callable that this node corresponds to, if any. */
+    DataFlowCallable asCallable() { result.getUnderlyingCallable() = this.asElement() }
+
+    /** Gets the target of this call, if any. */
+    Element getCallTarget() { result = this.asCall().getStaticCallTarget().getUnderlyingCallable() }
+
+    /** Gets a textual representation of this node. */
+    string toString() {
+      result = this.asElement().toString()
+      or
+      result = this.asNode().toString()
+      or
+      result = this.asCall().toString()
+    }
+
+    /** Gets the location of this node. */
+    Location getLocation() {
+      result = this.asElement().getLocation()
+      or
+      result = this.asNode().getLocation()
+      or
+      result = this.asCall().getLocation()
+    }
+  }
+
+  /** Provides additional sink specification logic. */
+  bindingset[c]
+  predicate interpretOutput(string c, InterpretNode mid, InterpretNode node) {
+    // Allow variables to be picked as output nodes.
+    exists(Node n, Element ast |
+      n = node.asNode() and
+      ast = mid.asElement()
+    |
+      c = "" and
+      n.asExpr().(VariableAccess).getTarget() = ast
+    )
+    // TODO: source fields?
```

Comment:
I'm not convinced that we want specific fields logic here. Do we have a use-case for this?

Reply:
Yeah, I think not.  Fields can already be selected as a source / sink via by specifying the class + "variable" (field) name.  This already works in tests.  Fields can also be selected in summary access paths, this also works in tests.  There are a couple of cases that aren't working:
- a summary expressing flow between fields of a global variable.  I've no idea why this isn't working.
- a source/sink that is a field of the qualifier to a call.  I think this is just because models-as-data doesn't support proper access paths on sources/sinks.

Reply:
Note to self: for "a summary expressing flow betwee fields of a global variable", the results could be blocked by `allowParameterReturnInSelf`???

Reply:
> the results could be blocked by `allowParameterReturnInSelf`???

Doesn't appear to be the case.